### DISCUSSION TITLE
Does View need tabIndexes?

### DISCUSSION THREAD

**Discussion by migueldeicaza** [tui-cs/Terminal.Gui#1080] [General]
Hello,

While reviewing the code for `View.cs`, it is not clear to me that `tabIndexes` is needed, and perhaps I am missing something subtle about it.

As far as I can tell, this value is allocated on demand when a new subview is added, but from that point on, `tabIndexes` contains a replica of the `subviews` list, and both of them are updated identically, except in the scenarios where the view is brought forward/backward.   

I found that this was introduced when this bug was opened: https://github.com/migueldeicaza/gui.cs/issues/742 and @BDisp submitted the fix.

Is the expectation that they would differ only in the presence of their placement on the hierarchy and this is why this is there?If not, it seems like every use of `tabIndexes` in addition, could be safely replaced with the equivalent operation on `subviews`.

**Comment by BDisp** [comment]
Before presenting the `tabIndexes`, I thought only using the `subviews `was sufficient, but then I faced with others issues:

1 - The `BringSubviewToFront`, `BringSubviewForward`, `SendSubviewToBack `and `SendSubviewBackwards `methods, change the display order but also the navigation order, which perhaps we don't want.

2 - Sometimes we need to create first a view because will be referenced by the next view, but we want to display after the the next view. So the navigation order is changed and if some of the above methods is used more complicate is for ordering the list.

3 - When interacting with the `subviews `to do the ordering tabindex, errors come up with the collection was changed, difficulting more the job.

4 - Due all this issues, I decided create the `tabIndexes `and opt to use the `subviews `to interact with layout/redraw and `tabIndexes `for navigation/focus.

Concluding, perhaps was possible to do that without `tabIndexes `but it was giving me a big headache.

**Comment by BDisp** [comment]
The `DynamicsMenuBar `scenario some views  use the `TabIndex `property which interact with the `tabIndexes`.

**Comment by migueldeicaza** [comment]
Thanks for the explanation, that makes sense.

You are right that currently the tab order is done via the order in a list (subviews originally, and tabIndexes now).   

One thing that some other toolkits have done in the past is to do spacial navigation (down tries to find the next view down, left the next view left), but it is tricky to get right and requires other idioms to be introduced - which would be more complex and not sure anyone would like to use.

Cheers!
